core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/messaginglayer/NetworkCommunicationActor.scala

```diff
@@ -83,9 +99,135 @@ class NetworkCommunicationActor(parentRef: ActorRef, val actorId: ActorVirtualId
     */
   var networkMessageID = 0L
 
+  // data structures needed by flow control
+  var nextSeqNumForMainActor = 0L // used to send backpressure enable/disable messages to parent
+  val flowControl = new FlowControl()
+
+  private def incrementBacklogIfDataMessage(
+      id: ActorVirtualIdentity,
+      msg: WorkflowMessage
+  ): Unit = {
+    if (msg.msgType == WorkflowMessageType.DATA_MESSAGE) {
+      flowControl.receiverIdToDataBacklog(id) =
+        flowControl.receiverIdToDataBacklog.getOrElseUpdate(id, 0) + 1
+    }
+  }
+
+  private def decrementBacklogIfDataMessage(
+      receiverId: ActorVirtualIdentity,
+      msg: WorkflowMessage
+  ): Unit = {
+    if (msg.msgType == WorkflowMessageType.DATA_MESSAGE) {
+      if (
+        flowControl.receiverIdToDataBacklog
+          .contains(receiverId) && flowControl.receiverIdToDataBacklog(receiverId) > 0
+      ) {
+        flowControl.receiverIdToDataBacklog(receiverId) =
+          flowControl.receiverIdToDataBacklog(receiverId) - 1
+      }
+    }
+  }
+
+  private def decrementCreditIfDataMessage(
+      receiverId: ActorVirtualIdentity,
+      msg: WorkflowMessage
+  ): Unit = {
+    if (msg.msgType == WorkflowMessageType.DATA_MESSAGE) {
+      flowControl.receiverIdToCredits(receiverId) =
+        flowControl.receiverIdToCredits.getOrElseUpdate(
+          receiverId,
+          Constants.unprocessedBatchesCreditLimitPerSender
+        ) - 1
+    }
+  }
+
+  /**
+    * The credits received as a result of NetworkAck() are updated in the data structures. Also if credit is 0.
+    * a regular polling service needs to be started to get credit periodically from receiver.
+    */
+  private def updateCredits(
+      receiverId: ActorVirtualIdentity,
+      credits: Int
+  ): Unit = {
+    if (credits <= 0) {
```

Comment:
can credit be less than 0? if not, I suggest we just throw exception if < 0

Reply:
discussed